common/src/balance_unit.rs

```diff
@@ -0,0 +1,570 @@
+            _ => (self.inner, other.inner),
+        };
+        left.cmp(&right)
+    }
+}
+
+impl PartialOrd for BalanceUnit {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(&other))
+    }
+}
+
+impl Default for BalanceUnit {
+    fn default() -> Self {
+        Self::zero()
+    }
+}
+
+impl Zero for BalanceUnit {
+    fn zero() -> Self {
+        // `is_divisible` = false here because if this zero value will have some operations with divisible asset, `is_divisible` is changed to true
+        Self::new(0, false)
+    }
+
+    fn is_zero(&self) -> bool {
+        self.inner.is_zero()
+    }
+}
+
+#[cfg(feature = "std")]
+impl Display for BalanceUnit {
+    fn fmt(&self, f: &mut Formatter<'_>) -> sp_std::fmt::Result {
+        let s = if self.is_divisible {
+            FixedWrapper::from(self.inner).get().unwrap().to_string()
+        } else {
+            self.inner.to_string()
+        };
+        write!(f, "{}", s)
+    }
+}
+
+impl BalanceUnit {
+    pub fn new(balance: Balance, is_divisible: bool) -> Self {
+        Self {
+            inner: balance,
+            is_divisible,
+        }
+    }
+
+    pub fn divisible(balance: Balance) -> Self {
+        Self::new(balance, true)
+    }
+
+    pub fn indivisible(balance: Balance) -> Self {
+        Self::new(balance, false)
+    }
+
+    pub fn copy_divisibility(&self, balance: Balance) -> Self {
+        Self::new(balance, self.is_divisible)
+    }
+
+    pub fn get(&self) -> &Balance {
+        &self.inner
+    }
+
+    pub fn set(&mut self, value: Balance) {
+        self.inner = value
+    }
+
+    pub fn is_divisible(&self) -> bool {
+        self.is_divisible
+    }
+
+    pub fn pow(&self, x: u32) -> Result<Self, ArithmeticError> {
+        let balance = if self.is_divisible {
+            FixedWrapper::from(self.inner).pow(x).try_into_balance()?
+        } else {
+            self.inner.checked_pow(x).ok_or(ArithmeticError::Overflow)?
+        };
+        Ok(Self::new(balance, self.is_divisible))
+    }
+
+    pub fn sqrt(&self) -> Result<Self, ArithmeticError> {
+        let balance = if self.is_divisible {
+            FixedWrapper::from(self.inner)
+                .sqrt_accurate()
+                .try_into_balance()?
+        } else {
+            self.inner.integer_sqrt()
+        };
+        Ok(Self::new(balance, self.is_divisible))
+    }
+}
+
+impl From<Balance> for BalanceUnit {
+    fn from(balance: Balance) -> Self {
+        Self {
+            inner: balance,
+            is_divisible: true,
+        }
+    }
+}
+
+impl Add for BalanceUnit {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        let (left, right) = match (self.is_divisible, rhs.is_divisible) {
+            (false, true) => (balance!(self.inner), rhs.inner),
+            (true, false) => (self.inner, balance!(rhs.inner)),
+            _ => (self.inner, rhs.inner),
+        };
+        let balance = left.saturating_add(right);
```

Comment:
I prefer to use regular `add` that can panic, but @vovac12 wanted to make it more safely and we agreed on `saturating` operations

Reply:
Why `saturating` and not checked? @wer1st @vovac12

Reply:
It's just not obvious for me, so I will be glad if you clarify :)

Reply:
It is not expected that balance can overflow. Here we use `saturating` to prevent panic. `checked` provides `Option` result that is not convenient for this case

Reply:
Hypothetically one can create an extremely cheap token that can reasonably to overflow, use it in some pallet, trigger saturation, and cause some unexpected behaviour in the pallet. This unexpected behaviour can cause unexpected troubles.

After all, it is a structure in `common` that should potentially be used across whole blockchain.

Reply:
Give an example of your case. Because there is no problem with extremely cheap asset. Do you mean extremely expensive probably?

Reply:
Btw, we cannot keep/display/process values more than u128::Max

Reply:
> Give an example of your case. Because there is no problem with extremely cheap asset. Do you mean extremely expensive probably?

As far as I understand, users can create arbitrary assets on the blockchain with `assets.register` extrinsic. It automatically gives permission to use `assets.mint` on the new asset for the author. Therefore they can mint up to u128::Max. Since the asset is unrestrictedly minted by the author it becomes basically worthless.

Reply:
It looks ok that user cannot mint more than u128::Max